Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs

```diff
@@ -4,103 +4,132 @@
-                var inputActionsAsset = objects.FirstOrDefault(o => o != null && o.name == kAssetName) as InputActionAsset;
-                if (inputActionsAsset != null)
-                    return inputActionsAsset;
+                return null;
             }
 
-            return CreateNewActionAsset();
+            var asset = ScriptableObject.CreateInstance<InputActionAsset>();
+            asset.name = InputSystem.kProjectWideActionsAssetName;
+            try
+            {
+                asset.LoadFromJson(text);
+            }
+            catch (Exception exception)
+            {
+                Debug.LogError($"Could not parse input actions in JSON format from '{kAssetPath}' ({exception})");
+                Object.DestroyImmediate(asset);
+                return null;
+            }
+            return asset;
         }
 
-        private static InputActionAsset CreateNewActionAsset()
+        internal static InputActionAsset LoadFromInputManagerSettings()
         {
-            var json = File.ReadAllText(Path.Combine(Environment.CurrentDirectory, s_DefaultAssetPath));
+            var objects = AssetDatabase.LoadAllAssetsAtPath(kInputManagerAssetPath);
+            if (objects != null)
+                return objects.FirstOrDefault(o => o != null && o.name == InputSystem.kProjectWideActionsAssetName) as InputActionAsset;
+            return null;
+        }
 
-            var asset = ScriptableObject.CreateInstance<InputActionAsset>();
-            asset.LoadFromJson(json);
-            asset.name = kAssetName;
+        internal static InputActionAsset GetOrCreate()
+        {
+#if UNITY_INCLUDE_TESTS
+            if (s_TestAsset != null) return s_TestAsset;
+#endif
 
-            AssetDatabase.AddObjectToAsset(asset, s_AssetPath);
+            InputActionAsset asset;
 
-            // Make sure all the elements in the asset have GUIDs and that they are indeed unique.
-            var maps = asset.actionMaps;
-            foreach (var map in maps)
-            {
-                // Make sure action map has GUID.
-                if (string.IsNullOrEmpty(map.m_Id) || asset.actionMaps.Count(x => x.m_Id == map.m_Id) > 1)
-                    map.GenerateId();
+            asset = LoadFromProjectSettings();
+            if (asset != null) return asset;
 
-                // Make sure all actions have GUIDs.
-                foreach (var action in map.actions)
-                {
-                    var actionId = action.m_Id;
-                    if (string.IsNullOrEmpty(actionId) || asset.actionMaps.Sum(m => m.actions.Count(a => a.m_Id == actionId)) > 1)
-                        action.GenerateId();
-                }
+            // v1.8.0-pre1 stored the Actions in ProjectSettings/InputManager.asset.
+            // Check if we have a project that was saved on that version and migrate it
+            // to save in the default location used from the pre2 release.
+            asset = LoadFromInputManagerSettings();
```

Comment:
I cannot see that we actually remove InputActionAsset from InputManager.asset here so are we really migrating and not just making a copy? I think we need to remove the old one from InputManager.asset, otherwise it would be enumerated and possible to load if loading InputManager.asset.